### PR TITLE
enable builds with PR references

### PR DESCRIPTION
**Pull request by gabemontero**
For https://trello.com/c/dY8fPwNo/1194-3-support-fetching-out-pr-ref-branches-techdebt

@openshift/devex PTAL

### CODE REVIEW

**Comment by gabemontero** [comment]
[(3) Support fetching out PR ref branches [techdebt]](https://trello.com/c/dY8fPwNo/1194-3-support-fetching-out-pr-ref-branches-techdebt)

**Comment by gabemontero** [comment]
[testextended][extended:core(should start a build from a PR ref)]

**Comment by gabemontero** [comment]
[test]

File: pkg/build/builder/source.go [hunk 1]
```diff
@@ -252,7 +252,16 @@ func extractGitSource(gitClient GitClient, gitSource *api.GitBuildSource, revisi
 		}
 
 		if err := gitClient.Checkout(dir, commit); err != nil {
-			return true, err
+			//TODO - do we want to parse the error string for the text that accompanies a clone / checkout of a PR ref, such as:   pathspec 'refs/pull/1/head' did not match any file(s) known to git
```
Comment (bparees) [inline pkg/build/builder/source.go]: i think this case is infrequent enough that it's not worth introducing the fragility that would come w/ that optimization.

File: pkg/build/builder/source.go [hunk 2]
```diff
@@ -252,7 +252,16 @@ func extractGitSource(gitClient GitClient, gitSource *api.GitBuildSource, revisi
...
+			err = gitClient.Fetch(dir, gitSource.URI, commit)
```
Comment (bparees) [inline pkg/build/builder/source.go]: it's probably worth adding some level 4 logging to this code path so we know when we've gone down it successfully (or if we ever go down it incorrectly).

File: pkg/build/builder/source.go (hunk 1 above)
Comment (gabemontero) [inline pkg/build/builder/source.go]: sounds good ... I'll remove the TODO comment

File: test/extended/builds/start.go [hunk 3]
```diff
@@ -51,6 +52,38 @@ var _ = g.Describe("[builds][Slow] starting a build using CLI", func() {
 		})
 	})
 
+	g.Describe("oc start-build with pr ref", func() {
+		g.It("should start a build from a PR ref, wait for the build to complete, and confirm the right level was used", func() {
+			g.By("make sure wildly is has latest tag")
```
Comment (bparees) [inline test/extended/builds/start.go]: s/is/imagestream/ (i read "is" as the verb..it was very confusing :) )

File: test/extended/builds/start.go [hunk 4]
```diff
@@ -51,6 +52,38 @@ var _ = g.Describe("[builds][Slow] starting a build using CLI", func() {
...
+			n := 0
+			for n < 100 {
+				is, _ := oc.AsAdmin().Run("get").Args("is", "wildfly", "-n", "openshift").Output()
+				n++
+				if strings.Contains(is, "latest") {
+					break
+				}
+				time.Sleep(100 * time.Millisecond)
+			}
```
Comment (bparees) [inline test/extended/builds/start.go]: wait.Poll()

**Comment by gabemontero** [comment]
On the 1 test failure, we hit flake https://github.com/openshift/origin/issues/12072

File: test/extended/builds/start.go (hunk 4 above)
Comment (gabemontero) [inline test/extended/builds/start.go]: Turns out there already is a helper function for this ... will switch to that.

File: test/extended/builds/start.go (hunk 3 above)
Comment (gabemontero) [inline test/extended/builds/start.go]: part of next push

**Comment by gabemontero** [comment]
updates pushed @bparees  - thanks

File: pkg/build/builder/source.go [hunk 5]
```diff
@@ -252,7 +252,17 @@ func extractGitSource(gitClient GitClient, gitSource *api.GitBuildSource, revisi
 		}
 
 		if err := gitClient.Checkout(dir, commit); err != nil {
-			return true, err
+			glog.V(4).Infof("Checkout after clone failed for ref %s with error: %#v, attempting fetch", commit, err)
```
Comment (bparees) [inline pkg/build/builder/source.go]: s/%#v/%v/

**Comment by bparees** [comment]
one last nit and lgtm.

**Comment by bparees** [comment]
[merge]

**Comment by gabemontero** [comment]
second test run hit flake https://github.com/openshift/origin/issues/13650

File: pkg/build/builder/source.go (hunk 5 above)
Comment (gabemontero) [inline pkg/build/builder/source.go]: just making sure @bparees - you decided you were OK with this and you posted the merge comment ?

Comment (bparees) [inline pkg/build/builder/source.go]: no, apparently I forgot :)  merge comment deleted.

Comment (gabemontero) [inline pkg/build/builder/source.go]: update pushed

**Comment by gabemontero** [comment]
nit addressed

**Comment by smarterclayton** [comment]
Sweet, want this for our own CI

**Comment by bparees** [comment]
> Sweet, want this for our own CI

we did it just for you. :gift:

**Comment by jim-minter** [comment]
nice!

**Comment by gabemontero** [comment]
On the third test run, we hit flake https://github.com/openshift/origin/issues/13650 again, as well as a network/env hiccup during the network minimal test:

```
Error: Error downloading packages:
  Curl error (35): SSL connect error for https://mirrors.fedoraproject.org/metalink?repo=fedora-25&arch=x86_64 [Encountered end of file]
The command '/bin/sh -c dnf -y update && dnf -y install bind-utils findutils hostname iproute iputils less procps-ng tar which bridge-utils ethtool iptables-services openvswitch && dnf clean all' returned a non-zero code: 1
[ERROR] PID 13696: hack/dind-cluster.sh:276: `${DOCKER_CMD} build -t "${image_name}" .` exited with status 1.
```

@bparees - I think we are good for merge .... I saw some conversation in other PRs about whether it was valid to post the merge comment while things were on hold for the rebase, but I don't know if I saw a conclusion as to what was valid.  So please post/not post as needed.  Thanks.

**Comment by bparees** [comment]
[merge] (queue is disabled but will merge once it's back up)

**Pull request merged**